WikiWatershed/model-my-watershed
# Shippensburg DRB 2100 Land Analysis

## Overview

Adds analysis of Shippensburg 2100 Land Forecast layers. This is done by querying the Fast Zonal API developed by Drexel / ANS, investigated in #3323. The query is wrapped in the standard Celery API used for all other analyses, and is identical from the client perspective. This API is publicly available in the MMW Public API.

The analysis is DRB specific, and is only available in areas of interest within the DRB. This is the first analyze endpoint to have such a limitation, and is validated upon request, returning an HTTP 400 if the requirement is not met. The front-end only queries these endpoints for AOIs within the DRB. The DRB check is also standardized between the front- and back-ends.

Some layer name cleanup is also done, as requested.

### TODO

- [x] Figure out what to put in the **Source** line for the analysis (emailed Dave, waiting for response)
- [x] Ensure that the API description is correct (emailed Dave, waiting for response)
- [x] Ensure that, in the case of Drexel / ANS API failure, the other analyses (NLCD, Protected Lands) are not affected
- [x] Add all possible rasters to the allowed rasters list.

Connects #3325 
Connects #3332 
Connects #3328 

### Demo

<img width="1286" alt="image" src="https://user-images.githubusercontent.com/1430060/85216877-0c1de080-b358-11ea-9f40-51ed1914c265.png">

<img width="1286" alt="image" src="https://user-images.githubusercontent.com/1430060/85216941-cc0b2d80-b358-11ea-9461-050587bd3170.png">

## Testing Instructions

* Check out this branch and `bundle` and restart your Celery service to pick up new tasks `vagrant ssh worker -c 'sudo service celeryd restart'`
* Go to [:8000/](http://localhost:8000/) and pick a shape that is obviously in the DRB
  - [x] Ensure you see the two DRB analyses under the Land tab
* Change the shape, and choose something obviously not in the DRB
  - [x] Ensure you don't see those two options under Land analysis
* Change the shape, and choose something on the border, such as Upper Mahanoy Creek
  - [x] Ensure you don't see those two options under Land analysis
* Edit `geoprocessing_api/calcs.py` and add the following line:
    ```diff
    diff --git a/src/mmw/apps/geoprocessing_api/calcs.py b/src/mmw/apps/geoprocessing_api/calcs.py
    index c6163d3d..f88ed3a9 100644
    --- a/src/mmw/apps/geoprocessing_api/calcs.py
    +++ b/src/mmw/apps/geoprocessing_api/calcs.py
    @@ -340,6 +340,8 @@ def drexel_fast_zonal(geojson, dataset):
     
         res = requests.post(settings.DREXEL_FAST_ZONAL_API['url'], data=payload)
     
    +    raise Exception('boom')
    +
         # Raise any non-success responses as exceptions, indicating failure
         res.raise_for_status()
    ```
* Restart Celery to pick up the changes `vagrant ssh worker -c 'sudo service celeryd restart'`
* Pick a shape within the DRB
  - [ ] Ensure that while the DRB checks fail, the other two analyses render correctly
